nixos/modules/services/networking/radicale.nix

```diff
@@ -43,50 +54,138 @@ in
+        This option is mutually exclusive with <literal>config</literal>.
+      '';
+      example = literalExample ''
+        server = {
+          hosts = [ "0.0.0.0:5232" "[::]:5232" ];
+        };
+        auth = {
+          type = "htpasswd";
+          htpasswd_filename = "/etc/radicale/users";
+          htpasswd_encryption = "bcrypt";
+        };
+        storage = {
+          filesystem_folder = "/var/lib/radicale/collections";
+        };
+      '';
+    };
+
+    rights = mkOption {
+      type = format.type;
+      description = ''
+        Configuration for Radicale's rights file. See
+        <link xlink:href="https://radicale.org/3.0.html#documentation/authentication-and-rights" />.
+        This option only works in conjunction with <literal>settings</literal>.
+        Any value other than <literal>{ }</literal> will set
+        <literal>settings.rights.type</literal> and
+        <literal>settings.rights.file</literal> to approriate values.
+      '';
+      default = { };
+      example = literalExample ''
+        root = {
+          user = ".+";
+          collection = "";
+          permissions = "R";
+        };
+        principal = {
+          user = ".+";
+          collection = "{user}";
+          permissions = "RW";
+        };
+        calendars = {
+          user = ".+";
+          collection = "{user}/[^/]+";
+          permissions = "rw";
+        };
+      '';
+    };
+
+    extraArgs = mkOption {
       type = types.listOf types.str;
       default = [];
       description = "Extra arguments passed to the Radicale daemon.";
     };
   };
 
   config = mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = cfg.settings == { } || cfg.config == "";
+        message = ''
+          The options services.radicale.config and services.radicale.settings
+          are mutually exclusive.
+        '';
+      }
+    ];
+
+    services.radicale.settings.rights = mkIf (cfg.rights != { }) {
+      type = "from_file";
+      file = toString rightsFile;
+    };
+
     environment.systemPackages = [ cfg.package ];
 
     users.users.radicale =
       { uid = config.ids.uids.radicale;
         description = "radicale user";
         home = "/var/lib/radicale";
-        createHome = true;
       };
 
     users.groups.radicale =
       { gid = config.ids.gids.radicale; };
 
+    # The service configuration should resemble
+    # https://radicale.org/3.0.html#tutorials/running-as-a-service/linux-with-systemd-system-wide
     systemd.services.radicale = {
       description = "A Simple Calendar and Contact Server";
       after = [ "network.target" ];
+      requires = [ "network.target" ];
       wantedBy = [ "multi-user.target" ];
       serviceConfig = {
         ExecStart = concatStringsSep " " ([
           "${cfg.package}/bin/radicale" "-C" confFile
         ] ++ (
           map escapeShellArg cfg.extraArgs
         ));
+        Restart = "on-failure";
```

Comment:
Why? Failures in configuration should become obvious. Restarting a software that fails is like ignoring that it segfaults 50% of the time during startup. It is a bug that must be fixed.

Reply:
I just copied that from upstream.
But I agree, so I removed it.

Reply:
Aren't many services doing this? I mean if it crashes it probably should restart instead of being down.

Reply:
I agree. The proper solution here is to properly configure system monitoring, using the `systemd` failure notification mechanisms, etc... I don't like finding out a service is down the next time I use it.

Reply:
> Aren't many services doing this?

A few, but it's a minority I think.

> The proper solution here is to properly configure system monitoring, using the systemd failure notification mechanisms, etc... I don't like finding out a service is down the next time I use it.

Once you configure that you can also set `Restart = "on-failure"` but the default should, like @andir said, make broken configurations obvious.

Reply:
As a general point his really assumes you run your OS as a pet and not cattle... though specifically for this service I guess that is fine because if you run this service it's almost certainly for personal use on a pet machine, so I'll mark this as resolved.